include/cuda/std/detail/libcxx/include/bit

```diff
@@ -53,58 +53,300 @@ namespace std {
+#pragma GCC system_header
+#endif
+
+_LIBCUDACXX_PUSH_MACROS
+#include <__undef_macros>
+
+#endif // __cuda_std__
+
 #if defined(__IBMCPP__)
 #include "support/ibm/support.h"
 #endif
 #if defined(_LIBCUDACXX_COMPILER_MSVC)
 #include <intrin.h>
 #endif
 
-#if defined(_LIBCUDACXX_USE_PRAGMA_GCC_SYSTEM_HEADER)
-#pragma GCC system_header
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+#define _LIBCUDACXX_BIT_CONSTEXPR constexpr
+
+#ifndef _LIBCUDACXX_COMPILER_MSVC
+
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __binary_ctz2(uint64_t __x, int __c)           _NOEXCEPT {
+  return (__x & 0x1) ? __c : __c + 1;
+}
+
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __binary_ctz4(uint64_t __x, int __c)           _NOEXCEPT {
+  return __binary_ctz2(
+            __x >> 2*!(__x & 0x3),
+            __c + 2*!(__x & 0x3));
+}
+
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __binary_ctz8(uint64_t __x, int __c)           _NOEXCEPT {
+  return __binary_ctz4(
+            __x >> 4*!(__x & 0x0F),
+            __c + 4*!(__x & 0x0F));
+}
+
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __binary_ctz16(uint64_t __x, int __c)           _NOEXCEPT {
+  return __binary_ctz8(
+            __x >> 8*!(__x & 0x00FF),
+            __c + 8*!(__x & 0x00FF));
+}
+
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __binary_ctz32(uint64_t __x, int __c)           _NOEXCEPT {
+  return __binary_ctz16(
+            __x >> 16*!(__x & 0x0000FFFF),
+            __c + 16*!(__x & 0x0000FFFF));
+}
+
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __binary_ctz64(uint64_t __x)           _NOEXCEPT {
+  return __binary_ctz32(
+            __x >> 32*!(__x & 0x00000000FFFFFFFF),
+            32*!(__x & 0x00000000FFFFFFFF));
+}
+
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __libcpp_ctz(unsigned __x)           _NOEXCEPT {
+#if defined(__CUDA_ARCH__)
+#  if (_LIBCUDACXX_STD_VER >= 14)
+  if (__libcpp_is_constant_evaluated()) {
+    return __binary_ctz32(static_cast<uint64_t>(__x), 0);
+  }
+  else {
+    return (!__x) ? 32 : __ffs(__x) - 1;
+  }
+#  else
+  return __binary_ctz32(static_cast<uint64_t>(__x), 0);
+#  endif
+#else
+  return __builtin_ctz(__x);
 #endif
+}
 
-_LIBCUDACXX_PUSH_MACROS
-#include <__undef_macros>
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __libcpp_ctz(unsigned long __x)      _NOEXCEPT {
+#if defined(__CUDA_ARCH__)
+#  if (_LIBCUDACXX_STD_VER >= 14)
+  if (__libcpp_is_constant_evaluated()) {
+    return __binary_ctz64(static_cast<uint64_t>(__x));
+  }
+  else {
+    return (!__x) ? 32 : __ffsll(__x) - 1;
+  }
+#  else
+  return __binary_ctz64(static_cast<uint64_t>(__x));
+#  endif
+#else
+  return __builtin_ctzl(__x);
+#endif
+}
 
-_LIBCUDACXX_BEGIN_NAMESPACE_STD
+inline _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_BIT_CONSTEXPR
+int __libcpp_ctz(unsigned long long __x) _NOEXCEPT {
+#if defined(__CUDA_ARCH__)
```

Comment:
Should these be if target instead of checks for `__CUDA_ARCH__`?

Reply:
This is behavior specific to NVCC. (Also NV_IF_TARGET embeds brackets due to if target requiring them `{...}` and this breaks C++11 constexpr.